lib/appoptics_apm/inst/grpc_client.rb

```diff
@@ -0,0 +1,161 @@
+# Copyright (c) 2018 SolarWinds, LLC.
+# All rights reserved.
+
+module AppOpticsAPM
+  module GRPC
+
+    module ActiveCall
+      if defined? ::GRPC
+        StatusCodes = {}
+        ::GRPC::Core::StatusCodes.constants.each { |code| StatusCodes[::GRPC::Core::StatusCodes.const_get(code)] = code }
+      end
+
+      def self.included(klass)
+        ::AppOpticsAPM::Util.method_alias(klass, :request_response, ::GRPC::ActiveCall)
+        ::AppOpticsAPM::Util.method_alias(klass, :client_streamer, ::GRPC::ActiveCall)
+        ::AppOpticsAPM::Util.method_alias(klass, :server_streamer, ::GRPC::ActiveCall)
+        ::AppOpticsAPM::Util.method_alias(klass, :bidi_streamer, ::GRPC::ActiveCall)
+      end
+
+      def grpc_tags(method_type, method)
+        tags = { 'Spec' => 'rsc',
+                 'RemoteURL' => "grpc://#{peer}#{method}",
+                 'GRPCMethodType' => method_type,
+                 'IsService' => 'True'
+        }
+        tags['Backtrace'] = AppOpticsAPM::API.backtrace if AppOpticsAPM::Config[:grpc_client][:collect_backtraces]
+        tags
+      end
+
+      def request_response_with_appoptics(req, metadata: {})
+        unary_response(req, type: 'UNARY', metadata: metadata, without: :request_response_without_appoptics)
+      end
+
+      def client_streamer_with_appoptics(req, metadata: {})
+        unary_response(req, type: 'CLIENT_STREAMING', metadata: metadata, without: :client_streamer_without_appoptics)
+      end
+
+      def server_streamer_with_appoptics(req, metadata: {}, &blk)
+        @tags = grpc_tags('SERVER_STREAMING', metadata[:method] || metadata_to_send[:method])
+        AppOpticsAPM::API.log_entry('grpc_client', @tags)
```

Comment:
I don't understand why some `AppOpticsAPM` scopes are prefixed with `::` while others aren't. I don't see nested `AppOpticsAPM` scopes so it seems like either one would work but don't understand why there are both.

Reply:
All them could but some may have to. It was all mixed up when I started and I never got around to figuring out where the `::` prefix would actually be required. I'll create a Jira for a future cleanup.

Reply:
i have plenty of jira's for cleanup too.